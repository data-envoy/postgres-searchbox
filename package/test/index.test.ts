import pkg from 'pg';
const { Client } = pkg;
import format from 'pg-format';
// Scripts
import { initTestDatabase } from '@scripts/mock-data.js';
import { createColumnAndIndex } from '@scripts/create-index.js';
// Main functions
import { searchHandler } from '@/index.js';

/**
 * This file is for testing the server side functions
 * Although /scripts are imported here thay are not being tested
 * they're used to set the database state for the tests
 */

describe('requestHandler', () => {
  const tableName = 'test_table';
  const initTestDatabaseParams = { tableName, rowCount: 100, fakerSeed: 123 };

  const client = new Client();
  client.connect();

  beforeEach(async () => {
    // Drop the table so that the tests can be run independently
    const dropSql = format(`DROP TABLE IF EXISTS %I`, tableName);
    await client.query(dropSql);
  });

  afterAll(async () => {
    const dropSql = format('DROP TABLE IF EXISTS %I', tableName);
    await client.query(dropSql);
    await client.end();
  });

  /**
   * Test handler returns results
   */

  it('should return results', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const query = 'affordable keyboard';

    const req = {
      body: JSON.stringify({ params: { query }, indexName: tableName }),
    };
    const res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn().mockReturnThis(),
    };
    await searchHandler(req, res);

    // Test status and json are called
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.json).toHaveBeenCalledWith({
      query,
      results: [
        {
          hits: expect.any(Array),
          hitsPerPage: 20,
          page: 0,
          nbHits: 1,
          nbPages: 1,
        },
      ],
    });

    const expectedResult = {
      id: 4,
      name: 'Bespoke Cotton Keyboard',
      description:
        'The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design',
      price: 8026,
    };
    // Test hits contained expected object
    expect(res.json.mock.calls[0][0].results[0].hits).toContainEqual(
      expectedResult
    );
  });

  it('should return multiple pages', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn().mockReturnThis(),
    };

    /**
     * Page 1
     */

    const params_1 = { query: 'good', page: 0 };
    const req_1 = { body: { params: params_1, indexName: tableName } };
    await searchHandler(req_1, res);

    // Test status and json are called
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.json).toHaveBeenCalledWith({
      query: params_1.query,
      results: [
        {
          hits: expect.any(Array),
          hitsPerPage: 20,
          page: 0,
          nbHits: 23,
          nbPages: 2,
        },
      ],
    });

    /**
     * Page 2
     */

    const params_2 = { query: 'good', page: 1 };
    const req_2 = { body: { params: params_2, indexName: tableName } };
    await searchHandler(req_2, res);

    expect(res.json).toHaveBeenCalledWith({
      query: params_1.query,
      results: [
        {
          hits: expect.any(Array),
          hitsPerPage: 20,
          page: 1,
          nbHits: 23,
          nbPages: 2,
        },
      ],
    });
  }, 10_000);

  /**
   * Test page that's out of range
   */

  it('should return empty array', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn().mockReturnThis(),
    };

    const req = {
      body: { params: { query: 'good', page: 3 }, indexName: tableName },
    };
    await searchHandler(req, res);

    // Test status and json are called
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.json).toHaveBeenCalledWith({
      query: req.body.params.query,
      results: [
        {
          hits: [],
          hitsPerPage: 20,
          page: 3,
          nbHits: 23,
          nbPages: 2,
        },
      ],
    });
  });

  /**
   * Test sorting by column
   */

  it('should sort by column', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn().mockReturnThis(),
    };

    const req = {
      body: {
        params: { query: 'good', page: 0 },
        indexName: `${tableName}?sort=price+asc,name+asc`,
      },
    };
    await searchHandler(req, res);

    // Test status and json are called
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.json).toHaveBeenCalledWith({
      query: req.body.params.query,
      results: [
        {
          hits: expect.any(Array),
          hitsPerPage: 20,
          page: 0,
          nbHits: 23,
          nbPages: 2,
        },
      ],
    });

    // Test hits are sorted by price
    const hits = res.json.mock.calls[0][0].results[0].hits;
    const prices: number[] = hits.map((hit: { price: number }) => hit.price);
    expect(prices).toEqual(prices.sort((a, b) => a - b));
  });

  it('should sort by column desc', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const res = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn().mockReturnThis(),
    };

    const req = {
      body: {
        params: { query: 'good', page: 0 },
        indexName: `${tableName}?sort=price+desc`,
      },
    };
    await searchHandler(req, res);

    // Test status and json are called
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.json).toHaveBeenCalledWith({
      query: req.body.params.query,
      results: [
        {
          hits: expect.any(Array),
          hitsPerPage: 20,
          page: 0,
          nbHits: 23,
          nbPages: 2,
        },
      ],
    });

    // Test hits are sorted by price
    const hits = res.json.mock.calls[0][0].results[0].hits;
    const prices: number[] = hits.map((hit: { price: number }) => hit.price);
    expect([...prices]).toEqual(prices.sort((a, b) => b - a));
  });
});
