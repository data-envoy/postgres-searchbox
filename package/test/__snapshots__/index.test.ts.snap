// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestHandler should handle highlighting 1`] = `
{
  "_highlightResult": {
    "description": {
      "fullyHighlighted": false,
      "matchLevel": "full",
      "matchedWords": [],
      "value": "The slim & simple Maple Gaming __ais-highlight__Keyboard__/ais-highlight__ from Dev Byte comes with a sleek body and",
    },
    "name": {
      "fullyHighlighted": false,
      "matchLevel": "full",
      "matchedWords": [],
      "value": "__ais-highlight__Bespoke__/ais-highlight__ Granite __ais-highlight__Keyboard__/ais-highlight__",
    },
  },
  "brand": "Powlowski, Brekke and Hamill",
  "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
  "id": 20,
  "name": "Bespoke Granite Keyboard",
  "price": 14,
}
`;

exports[`requestHandler should return facets on request with empty query 1`] = `
{
  "brand": {
    "Abshire - Carroll": 2,
    "Hammes Group": 2,
    "Hayes - Kutch": 2,
    "Hickle - Bins": 2,
    "Kessler Inc": 2,
    "Kiehn Inc": 2,
    "Langworth, Kuvalis and Kohler": 2,
    "Littel, Borer and Bergstrom": 2,
    "Muller - Sporer": 2,
    "Smitham Group": 2,
  },
  "price": {
    "11": 6,
    "12": 7,
    "13": 10,
    "14": 6,
    "19": 5,
    "22": 8,
    "24": 9,
    "33": 7,
    "38": 7,
    "39": 6,
  },
}
`;

exports[`requestHandler should return results 1`] = `
{
  "brand": "Powlowski, Brekke and Hamill",
  "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
  "id": 20,
  "name": "Bespoke Granite Keyboard",
  "price": 14,
}
`;
